File: src/java/PhantasyIslandJavaRemoteControl/AirplaneManager.java
package PhantasyIslandJavaRemoteControl;

import java.util.HashMap;
import java.util.Map;

/**
 * 管理并更新所有飞机状态的管理器。
 */
public class AirplaneManager {
    private static final AirplaneManager instance = new AirplaneManager();
    public Map<String, Object> airplanesTable = new HashMap<>(); // 实际上是 AirplaneController

    private AirplaneManager() {}

    public static AirplaneManager getAirplaneManager() {
        return instance;
    }

    public Object ping() {
        // return HttpLayer.ping();
        return null;
    }

    public void start() {
        // HttpLayer.start();
    }

    public Object getAirplane(String id) {
        return airplanesTable.get(id);
    }

    public void flush() {
        // Map<String, Map<String, Object>> statusMap = HttpLayer.getAllAirplaneStatus();
        // 遍历并更新 airplanesTable
    }
}

File: src/kt/PhantasyIslandKotlinRemoteControl/AirplaneManager.kt
package PhantasyIslandKotlinRemoteControl

import org.json.JSONObject

class AirplaneManager {
    val airplanesTable: MutableMap<String, AirplaneController> = mutableMapOf()

    fun ping(): JSONObject? {
        return HttpLayer.ping()
    }

    fun pingVolatile(): JSONObject? {
        return HttpLayer.pingVolatile()
    }

    fun start(): JSONObject? {
        return HttpLayer.start()
    }

    fun startVolatile(): JSONObject? {
        return HttpLayer.startVolatile()
    }

    fun getAirplane(id: String): AirplaneController? {
        return airplanesTable[id]
    }

    fun sleep(timeMs: Long) {
        Thread.sleep(timeMs)
    }

    fun flush() {
        val allStatus = HttpLayer.getAllAirplaneStatus()
        val airplaneStatusMap = HttpLayer.processAirplane(allStatus)
        
        airplaneStatusMap?.forEach { (k, status) ->
            val existing = airplanesTable[k]
            val keyName = status["keyName"] as String
            val typeName = status["typeName"] as String
            val updateTimestamp = status["updateTimestamp"] as Long
            val statusObj = status["status"] as JSONObject
            val cameraFront = status["cameraFront"] as? String
            val cameraDown = status["cameraDown"] as? String
            
            val flyStatus = makeAirplaneFlyStatus(statusObj)

            if (existing == null) {
                airplanesTable[k] = AirplaneController(
                    keyName = keyName,
                    typeName = typeName,
                    updateTimestamp = updateTimestamp,
                    status = flyStatus,
                    cameraFront = cameraFront,
                    cameraDown = cameraDown
                )
            } else {
                existing.keyName = keyName
                existing.typeName = typeName
                existing.updateTimestamp = updateTimestamp
                existing.status = flyStatus
                existing.cameraFront = cameraFront
                existing.cameraDown = cameraDown
            }
        }
    }

    companion object {
        private val instance = AirplaneManager()
        
        @JvmStatic
        fun getAirplaneManager(): AirplaneManager {
            return instance
        }
    }
}
